compiler/rustc_lint/src/unused.rs

```diff
@@ -255,31 +350,46 @@ impl<'tcx> LateLintPass<'tcx> for UnusedResults {
                     // If the array is definitely non-empty, we can do `#[must_use]` checking.
                     Some(n) => {
                         let descr_pre = &format!("{}array{} of ", descr_pre, plural_suffix,);
-                        check_must_use_ty(cx, ty, expr, span, descr_pre, descr_post, n as usize + 1)
+                        check_must_use_ty(
+                            cx,
+                            ty,
+                            expr,
+                            span,
+                            descr_pre,
+                            descr_post,
+                            n as usize + 1,
+                            attr_to_check,
+                        )
                     }
                 },
-                ty::Closure(..) => {
-                    cx.struct_span_lint(UNUSED_MUST_USE, span, |lint| {
-                        let mut err = lint.build(&format!(
-                            "unused {}closure{}{} that must be used",
-                            descr_pre, plural_suffix, descr_post,
-                        ));
-                        err.note("closures are lazy and do nothing unless called");
-                        err.emit();
-                    });
-                    true
-                }
-                ty::Generator(..) => {
-                    cx.struct_span_lint(UNUSED_MUST_USE, span, |lint| {
-                        let mut err = lint.build(&format!(
-                            "unused {}generator{}{} that must be used",
-                            descr_pre, plural_suffix, descr_post,
-                        ));
-                        err.note("generators are lazy and do nothing unless resumed");
-                        err.emit();
-                    });
-                    true
-                }
+                ty::Closure(..) => match attr_to_check {
```

Comment:
I'd prefer a guard in the match arm: `if matches!(attr_to_check, WhatAttrToCheck::MustUse)`